build_test/test/written_asset_reader_test.dart

```diff
@@ -0,0 +1,43 @@
+import 'package:build/build.dart';
+import 'package:build_test/build_test.dart';
+import 'package:glob/glob.dart';
+import 'package:test/test.dart';
+
+void main() {
+  WrittenAssetReader reader;
+
+  setUp(() async {
+    final writer = InMemoryAssetWriter();
+    await writer.writeAsString(AssetId.parse('foo|a.txt'), 'a');
+    await writer.writeAsString(AssetId.parse('foo|lib/b.dart'), 'b');
+    await writer.writeAsString(AssetId.parse('bar|a.txt'), 'b_a');
+
+    reader = WrittenAssetReader(writer);
+  });
+
+  test("doesn't read assets that weren't written", () async {
+    final doesntExist = AssetId.parse('foo|c.txt');
+    expect(reader.canRead(doesntExist), completion(isFalse));
```

Comment:
nit: you can just use async/await here instead of relying on the async completers which are sort of weird (the rest of the test still executes before the expect is validated since there is no await).

`expect(await reader.canRead(doesntExist), false);`

Same for tests below (except the stream one)